transmart-core-db/grails-app/services/org/transmartproject/db/clinical/MultidimensionalDataResourceService.groovy

```diff
@@ -787,112 +746,44 @@ class MultidimensionalDataResourceService implements MultiDimensionalDataResourc
 
     /**
      * @description Function for getting a aggregate value of a single field.
-     * The allowed queryTypes are MIN, MAX and AVERAGE.
-     * The responsibility for checking the queryType is allocated to the controller.
+     *
+     * @param types set of aggregate functions to calculate
      * @param query
      * @param user
      */
-    @Override Map aggregate(List<AggregateType> types, MultiDimConstraint constraint_, User user) {
-        def constraint = (Constraint) constraint_
+    @Override Map<AggregateFunction, Number> aggregate(Set<AggregateFunction> types, MultiDimConstraint constraint, User user) {
+        assert constraint instanceof Constraint
         checkAccess(constraint, user)
 
-        def builder = getCheckedQueryBuilder(user)
-
-        def fieldTypes = types.collect { aggregateFieldType(it) }.unique()
-        if(fieldTypes.findAll().size() > 1) throw new InvalidQueryException(
-                "aggregate queries on numeric and textual values can not be combined in a single call")
-
-        if(fieldTypes.size() == 0) return [:]
-
-        def typedConstraint = fieldTypes.size() != 1 ? constraint :
-                new AndConstraint(args: [constraint, new FieldConstraint(
-                    operator: Operator.EQUALS,
-                    field: valueTypeField,
-                    value: fieldTypes[0],
-                )])
-
-        // get aggregate value
-        DetachedCriteria queryCriteria = builder.buildCriteria(typedConstraint)
-        def result = getAggregate(types, queryCriteria)
-
-        if (result == null || result.values().any {it == null || (it instanceof List && ((List) it).empty)}) {
-            // results not found, do some diagnosis to discover why not so we can return a useful error message
-            diagnoseEmptyAggregate(types, constraint, builder)
+        if (types.size() == 0) return [:]
 
-            // If diagnoseEmptyAggregate didn't throw any kind of exception, nothing left to do but to return the
-            // (empty) result
+        def builder = getCheckedQueryBuilder(user)
+        DetachedCriteria criteria = builder.buildCriteria(constraint)
+        def projections = Projections.projectionList()
+        types.each { AggregateFunction aggregateType ->
+            projections.add(projectionForAggregate(aggregateType), aggregateType.toString())
         }
-
-        return result
+        criteria
+                .setProjection(projections)
+                .setResultTransformer(Transformers.ALIAS_TO_ENTITY_MAP)
+                .add(Restrictions.eq('valueType', ObservationFact.TYPE_NUMBER))
```

Comment:
Also filter out null values?

Reply:
We are going to filter null values, but not null aggregate values. e.g. `{ stddev: null }` means there are less than two numerical values.